Mw/big upgrades

michaelwood opened the pull request.

Commit:
requirements: initial upgrade of requirements

Commit:
import: Comment out additional_data warning

Commit:
README: Add comment about new code

Commit:
urls: Upgrade all urls to use `path` instead of deprecated `url`

For upcoming django upgrade

Commit:
context_processors: Add `disable_cookie_popup` for testing/dev

The cookie popup steals focus which makes it difficult to test when the
cookie isn't stored. Timing for when it pops up is tricky to deal with
in the selenium tests.

Commit:
requirements_dev: Remove pytest, add python-black and upgrade everything

Commit:
tests: Rewrite and improve all existing tests

Refactor and improve tests:
* Upgrade to new selenium API (by_id -> By.ID)
* reverse_lazy in tests instead of string concat
* TestCase derived tests for better django compatibility and reducing
  the software stack/moving parts
* Improve timing for more reliable browser results
* Add Re-usable components for browser tests
* Split link checker from browser tests
* Remove tests which don't test anything
* Change json lines files to have correct file extension
* Update github workflow and split into new test links for parallel
  running

michaelwood commented:
Note to self last test fails are due to timing issues, checking for an item in "body" is pretty unreliable because selenium doesn't know what element to wait for in an implict wait similar problem to https://github.com/ThreeSixtyGiving/dataquality/issues/100

R2ZER0 commented on requirements.in:
```diff
@@ -1,4 +1,4 @@
-Django<3.3
+Django<6
```
Is all the grantnav code compatible with Django 5 too?

michaelwood replied:
Yeah I was in two mind about that, as I am aiming for the next LTS 5.2 that will be released,  4.x is the last available release for python3.8 which we will be upgrading soon so it won't upgrade us past 4.x right now.

https://docs.djangoproject.com/en/5.1/releases/5.0/#python-compatibility

michaelwood replied:
Have now set to <5 in https://github.com/ThreeSixtyGiving/grantnav/pull/1156/commits/0c40ad3ec1a5a44a00a5740c4af4fbddeb52f50c

R2ZER0 commented on README.md:
```diff
@@ -143,3 +143,6 @@ Adding and updating requirements
 
+## New Code
+
+New code should by python-black formatted
```
```suggestion
New code should be python-black formatted
```

R2ZER0 commented on tests/data/load_test_data.sh:
```diff
@@ -0,0 +1 @@
+./dataload/import_to_elasticsearch.py ./dataload/test_data/a002400000KeYdsAAF-currency.json ./dataload/test_data/a002400000OiDBQAA3.json ./dataload/test_data/awefw001p00000zgyHZAAY.json  --clean 
```
Running this command (or the script) gives FileNotFound error
```
FileNotFoundError: [Errno 2] No such file or directory: './dataload/test_data/a002400000KeYdsAAF-currency.json'
```
I'm guessing since they've all be moved

michaelwood replied:
I didn't mean to commit that file, thanks will remove it

Commit:
README: Fix typo in New code section

Co-authored-by: Rikki Guy <rikki.guy@opendataservices.coop>

Commit:
tests: Remove obsolete load_test_data script

This may be misleading as it only loads some of the test data that was
for a specific purpose.

Commit:
requirements: Keep Django <5 for now

This makes no change in the compiled requirements as it is not possible
to go higher than 4 as we're on python3.8 for now. However better to be
safer and not have any unintended upgrade here.

jarofgreen committed:
Add Coveralls

Only to test.yml, not test_links.yml

The latest version of the library was already included in requirements

https://github.com/ThreeSixtyGiving/grantnav/issues/1155

R2ZER0 commented on tests/browser_test_case.py:
```diff
@@ -0,0 +1,79 @@
+            profile.set_preference(
+                "browser.helperApps.neverAsk.saveToDisk", "application/json"
+            )
+            self.browser = getattr(webdriver, BROWSER)(firefox_profile=profile)
```
Running the tests with Firefox no longer works for me, with error:
```
Traceback (most recent call last):
  File "/home/rikki/Projects/grantnav/tests/browser_test_case.py", line 54, in setUp
    self.browser = getattr(webdriver, BROWSER)(firefox_profile=profile)
TypeError: __init__() got an unexpected keyword argument 'firefox_profile'
```
... but let's fix that later, just a "nice to have"

michaelwood replied:
That's odd because I haven't changed anything to do with that profile :\

R2ZER0 replied:
Probably just due to upgrades?

R2ZER0 commented on tests/browser_test_case.py:
```diff
@@ -0,0 +1,79 @@
+    def check_js_errors(self):
+        for log in self.browser.get_log("browser"):
+            # Datatables is sending warnings that we can't currently fix
+            # https://github.com/...
```
Is this meant to be a real link?

michaelwood replied:
Not sure, seems to be lost in time, I have raised a separate issue on this #1161 to investigate in the future.

R2ZER0 commented on tests/data/add_new_geo.py:
```diff
@@ -0,0 +1,38 @@
+import json
+import random
+
+#a002400000KeYdsAAF-currency.json  add_new_geo.py	       funders_old.jl				  README.md
```
Are these leftover comments?

Commit:
tests: browser_test_case: Remove link in comment about datatables

This seems to have been lost in time. Making a note:
Related: https://github.com/ThreeSixtyGiving/grantnav/issues/1161

Commit:
tests: data: Remove add_new_geo

This has served its purpose and no longer needed.

R2ZER0 reviewed (approved):
:+1:
